opengrok-indexer/src/main/java/org/opengrok/indexer/web/Util.java

```diff
@@ -677,13 +677,12 @@ public static void encode(String s, Appendable dest) throws IOException {
      * @param urlStr string URL
      * @return the encoded URL
      * @throws URISyntaxException URI syntax
-     * @throws MalformedURLException URL malformed
      */
-    public static String encodeURL(String urlStr) throws URISyntaxException, MalformedURLException {
-        URL url = new URL(urlStr);
-        URI constructed = new URI(url.getProtocol(), url.getUserInfo(),
-                url.getHost(), url.getPort(),
-                url.getPath(), url.getQuery(), url.getRef());
+    public static String encodeURL(String urlStr) throws URISyntaxException {
+        // URL url = new URL(urlStr); - this call
```

Comment:
I took a longer look at the departure from the `URL()` constructors yesterday. The `URL` class by itself does not support any encoding/decoding, it merely breaks down the URL into pieces. The encoding is then supplied by the `URI` class.

The problem is that `URI(String)` acceots only already encoded URIs and refuses any non-compliant characters. I think the way to go is to propagate the `URL` object all the way where possible, i.e. change the `String url` parameter type of the various `linkify()` and `buildLink()` methods to `URL url` and avoid passing the `URL` in the map of attributes - pass it as an extra argument with the `URL` type.

It is still not clear to me where the unescaped characters could originate from.